src/ContentValidationListener.php

```diff
@@ -356,6 +367,101 @@ protected function useRawData($controllerService)
         return false;
     }
 
+    /**
+     * @param $controllerService
+     * @return bool
+     */
+    protected function shouldRemoveEmptyData($controllerService)
+    {
+        if (! isset($this->config[$controllerService]['remove_empty_data'])
+            || (isset($this->config[$controllerService]['remove_empty_data'])
+                && $this->config[$controllerService]['remove_empty_data'] === true)
+        ) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * @param array $data Data to filter null values from
+     * @param array $compareTo Original data, send along to preserve keys/values in $data which are intentional
+     *
+     * @return array
+     */
+    // TODO strict types for PHP > 7 when module gets bumped
+    //    protected function removeEmptyData(array $data, array $compareTo = []) : array
+    protected function removeEmptyData($data, $compareTo = [])
+    {
+        // TODO when bumped to PHP > 7 and strict types above enforced this check may be removed
+        if (! is_array($data) || ! is_array($compareTo)) {
+            throw new TypeError(
+                sprintf(
+                    'Expected array\'s for function %s. Received "%s" and "%s".',
+                    __METHOD__,
+                    gettype($data),
+                    gettype($compareTo)
+                ),
+                500,
+                __FILE__
+            );
+        }
+
+        // Callback for array_filter() to remove null values (array_filter() removes 'false' values)
+        $removeNull = function ($value, $key = null) use ($compareTo) {
+            // If comparison array is empty, do a straight comparison
+            if (empty($compareTo)) {
+                return ! is_null($value);
+            }
+
+            // If key exists in comparison array, the 'null' value is on purpose, leave as is
+            if (array_key_exists($key, $compareTo)) {
+                return true;
+            }
+
+            return ! is_null($value);
+        };
+
+        $data = array_filter($data, $removeNull, ARRAY_FILTER_USE_BOTH);
+
+        if (empty($data)) {
+            return $data;
+        }
+
+        foreach ($data as $key => $value) {
+            if (! is_array($value)
+                && (! empty($value) || (is_bool($value)) && ! in_array($key, $compareTo))
+            ) {
+                continue;
+            }
+
+            if (! is_array($value)) {
+                unset($data[$key]);
+                continue;
+            }
+
+            if (empty(array_filter($value, $removeNull, ARRAY_FILTER_USE_BOTH))) {
+                unset($data[$key]);
+                continue;
+            }
+
+            if (array_key_exists($key, $compareTo) && is_array($compareTo[$key])) {
+                $tmpValue = $this->removeEmptyData($value, $compareTo[$key]);
+            } else {
+                $tmpValue = $this->removeEmptyData($value);
+            }
```

Comment:
This could become:

```php
$tmpValue = array_key_exists($key, $compareTo) && is_array($compareTo[$key])
    ? $this->removeEmptyData($value, $compareTo[$key])
    : $this->removeEmptyData($value);
```

Ternary assignments are nice for this kind of either/or thing, and make reading the code easier.

Reply:
done